### PR TITLE
Update metadata in existing assets when the model is modified

### LINKED ISSUE #277: Update metadata in existing assets when the model is modified
Updating an asset (or device) model metadata should automatically update corresponding assets:
 - add metadata: initialize with `null` or the default metadata value if any
 - delete metadata: remove the value

### PR DESCRIPTION
**Pull request by fmauNeko**
## What does this PR do ?
This PR implements the propagation of any metadata addition / deletion on an asset model to the linked assets.

Fixes #277

### CODE REVIEW

File: tests/scenario/modules/models/asset-model-metadata-propagation.test.ts [hunk 1]
```diff
@@ -0,0 +1,361 @@
...
+          width: { type: "integer" },
+        },
+        defaultValues: {
+          depth: 80,
+          width: 120,
+        },
+      },
+    });
+
+    await sdk.query<ApiAssetCreateRequest>({
```
Comment (OlivierCavadenti) [inline tests/scenario/modules/models/asset-model-metadata-propagation.test.ts:61]: put assets creation in fixtures file ?

File: lib/modules/model/ModelService.ts [hunk 2]
```diff
@@ -90,7 +96,32 @@ export class ModelService {
...
+        action: "search",
+        body: { query: { equals: { model } } },
+        controller: "device-manager/assets",
+        engineId: engine.index,
+        lang: "koncorde",
+      });
+
+      await Promise.all(
+        assets.result.hits.map((asset) => {
+          ask<AskAssetRefreshModel>("ask:device-manager:asset:refresh-model", {
```
Comment (OlivierCavadenti) [inline lib/modules/model/ModelService.ts:117]: ask is not awaited :'( :'(

**Review by OlivierCavadenti — CHANGES_REQUESTED** [review]

File: tests/scenario/modules/models/asset-model-metadata-propagation.test.ts (hunk 1 above)
Comment (fmauNeko) [reply tests/scenario/modules/models/asset-model-metadata-propagation.test.ts:61]: There's some variations in the assets between tests

File: lib/modules/asset/AssetService.ts [hunk 3]
```diff
@@ -52,6 +60,43 @@ export class AssetService {
...
+        for (const metadataName of Object.keys(
+          assetModel.asset.metadataMappings
+        )) {
+          modelMetadata[metadataName] = null;
+        }
+        for (const [metadataName, metadataValue] of Object.entries(
+          assetModel.asset.defaultMetadata
+        )) {
+          _.set(modelMetadata, metadataName, metadataValue);
+        }
```
[the remark this answers is not in the document]
Comment (fmauNeko) [reply lib/modules/asset/AssetService.ts:83]: Done !

File: lib/modules/asset/AssetService.ts [hunk 4]
```diff
@@ -67,34 +75,84 @@ export class AssetService {
...
+            const assetMetadata = Object.fromEntries(
+              Object.entries(asset._source.metadata).filter(([key]) => {
+                if (!(key in modelMetadata)) {
+                  removedMetadata.push(key);
+                  return false;
+                }
+
+                return true;
+              })
+            );
```
Comment (tdislay) [inline lib/modules/asset/AssetService.ts:130]: If the `delete` keyword is ok (in this case it seems to be)
Could you use a for loop like this:
```ts
for (const key of Object.keys(asset._source.metadata) {
  if (!(key in modelMetadata)) {
    removedMetadata.push(key);
    delete asset._source.metadata[key];
  }
}
```
suggested by @sebtiz13

**Review by tdislay — APPROVED** [review]

File: lib/modules/asset/AssetService.ts [hunk 5]
```diff
@@ -67,34 +75,84 @@ export class AssetService {
...
         }
 
-        const removedMetadata = _.difference(
-          Object.keys(asset._source.metadata),
-          Object.keys(modelMetadata)
+        const removedMetadata: string[] = [];
+
+        const updatedAssetsPerIndex: {
+          [x: string]: KDocument<AssetContent>[];
+        } = assetSearch.result.hits.reduce(
```
[the remark this answers is not in the document]
Comment (fmauNeko) [reply lib/modules/asset/AssetService.ts:116]: Probably none.

File: lib/modules/asset/AssetService.ts (hunk 4 above)
Comment (fmauNeko) [reply lib/modules/asset/AssetService.ts:130]: It should indeed be okay

Comment (fmauNeko) [reply lib/modules/asset/AssetService.ts:130]: It should indeed be okay !

File: lib/modules/asset/AssetService.ts [hunk 6]
```diff
@@ -52,6 +68,90 @@ export class AssetService {
...
     this.assetHistoryService = assetHistoryService;
+
+    this.registerAskEvents();
+  }
+
+  registerAskEvents() {
+    onAsk<AskAssetRefreshModel>(
+      "ask:device-manager:asset:refresh-model",
+      async ({ assetModel }) => {
+        const engines = await ask<AskEngineList>(
```
Comment (Aschen) [inline lib/modules/asset/AssetService.ts:79]: Can you declare a private method for clarity please?

File: lib/modules/asset/AssetService.ts [hunk 7]
```diff
@@ -52,6 +68,90 @@ export class AssetService {
...
+                engine.index,
+                [] as KDocument<AssetContent>[],
+              ])
+            )
+          );
+
+        for (const [index, updatedAssets] of Object.entries(
+          updatedAssetsPerIndex
+        )) {
+          await this.mReplace(null, index, updatedAssets, removedMetadata, {
```
Comment (Aschen) [inline lib/modules/asset/AssetService.ts:149]: You can parallelize those calls

File: lib/modules/asset/AssetService.ts [hunk 8]
```diff
@@ -306,6 +406,45 @@ export class AssetService {
     return result;
   }
 
+  /**
+   * Replace an asset metadata
+   */
+  public async mReplace(
```
Comment (Aschen) [inline lib/modules/asset/AssetService.ts:412]: Nit: from the behavior this function is replacing and historizing no?

**Review by Aschen — CHANGES_REQUESTED** [review]

File: lib/modules/asset/AssetService.ts (hunk 8 above)
Comment (fmauNeko) [reply lib/modules/asset/AssetService.ts:412]: That's what it does indeed

File: lib/modules/asset/AssetService.ts [hunk 9]
```diff
@@ -306,6 +406,45 @@ export class AssetService {
...
+    user: User,
+    engineId: string,
+    assets: KDocument<AssetContent>[],
+    removedMetadata: string[],
+    { refresh }: { refresh: any }
+  ): Promise<mReplaceResponse> {
+    const replacedAssets = await this.impersonatedSdk(
```
[the remark this answers is not in the document]
Comment (fmauNeko) [reply lib/modules/asset/AssetService.ts:419]: `user` is fixed at `null` on the call in the `Ask`. I kept the `user` parameter for consistency with the other SCRUD methods.

File: lib/modules/asset/AssetService.ts [hunk 10]
```diff
@@ -52,6 +68,15 @@ export class AssetService {
...
     this.config = plugin.config;
     this.assetHistoryService = assetHistoryService;
+
+    this.registerAskEvents();
+  }
+
+  registerAskEvents() {
+    onAsk<AskAssetRefreshModel>(
+      "ask:device-manager:asset:refresh-model",
+      this.refreshModel
```
Comment (Aschen) [inline lib/modules/asset/AssetService.ts:78]: The context is not bound for `refreshModel` (`this` will not be available inside the method)

File: lib/modules/asset/AssetService.ts (hunk 9 above)
[the remark this answers is not in the document]
Comment (Aschen) [reply lib/modules/asset/AssetService.ts:344]: I understand but here there is no need to use the ImpersonatedSDK for the reasons I gave in my previous message. This one can not really be considered as a SCRUD method I think

**Review by tdislay — APPROVED** [review]

**Review by Aschen — APPROVED** [review]

**Pull request merged**